test/pkcs12_format_test.c

```diff
@@ -0,0 +1,918 @@
+        p12bio = generate_p12(safes, NULL, filename); \
+    }
+
+# define end_pkcs12_with_mac(mac)  \
+        p12bio = generate_p12(safes, &mac, filename); \
+    }
+
+
+#define start_contentinfo  \
+    {  \
+        STACK_OF(PKCS12_SAFEBAG) *bags = NULL;
+
+#define end_contentinfo  \
+        if (bags && !TEST_true(PKCS12_add_safe(&safes, bags, -1, 0, NULL))) \
+            goto err; \
+        sk_PKCS12_SAFEBAG_pop_free(bags, PKCS12_SAFEBAG_free); \
+        bags = NULL; \
+    }
+
+#define end_contentinfo_encrypted(enc)  \
+        if (bags && !TEST_true(PKCS12_add_safe(&safes, bags, enc.nid, enc.iter, enc.pass))) \
+            goto err; \
+        sk_PKCS12_SAFEBAG_pop_free(bags, PKCS12_SAFEBAG_free); \
+        bags = NULL; \
+    }
+
+
+#define add_certbag(cert, attrs)  \
+    if (!TEST_true(generate_certbag(&bags, cert, attrs, sizeof(attrs)/sizeof(PKCS12_ATTR)))) \
+        goto err;
+
+
+#define add_keybag(key, attrs, enc)  \
+    if (!TEST_true(generate_keybag(&bags, key, attrs, sizeof(attrs)/sizeof(PKCS12_ATTR), &enc))) \
+        goto err;
+
+
+#define add_secretbag(secret_nid, secret, attrs)  \
+    if (!TEST_true(generate_secretbag(&bags, secret_nid, secret, attrs, sizeof(attrs)/sizeof(PKCS12_ATTR)))) \
+        goto err;
+
+
+/* TODO: Support safe contents? This should combine any bags created inside it */
+#define start_safecontents  \
+    {  \
+        STACK_OF(PKCS12_SAFEBAG) *contentbags = NULL;
+
+#define end_safecontents  \
+    }
+
+/* --------------------------------------------------------------------------
+ * PKCS12 checker macros
+ */
+
+#define start_check_pkcs12_with_mac(mac) \
+    {  \
+        STACK_OF(PKCS7) *safes = NULL; \
+        int safe_idx = 0;              \
+        PKCS12 *p12 = from_bio_p12(p12bio, &mac); \
+        if (!TEST_ptr(p12)) \
+            goto err; \
+        safes = PKCS12_unpack_authsafes(p12); \
+        if (!TEST_ptr(safes)) \
+            goto err;
+
+#define start_check_pkcs12  \
+    {  \
+        STACK_OF(PKCS7) *safes = NULL; \
+        int safe_idx = 0;              \
+        PKCS12 *p12 = from_bio_p12(p12bio, NULL); \
+        if (!TEST_ptr(p12)) \
+            goto err; \
+        safes = PKCS12_unpack_authsafes(p12); \
+        if (!TEST_ptr(safes)) \
+            goto err;
+
+#define start_check_pkcs12_file_with_mac(fname, mac) \
+    {  \
+        STACK_OF(PKCS7) *safes = NULL; \
+        int safe_idx = 0;              \
+        PKCS12 *p12 = read_p12(fname, &mac); \
+        if (!TEST_ptr(p12)) \
+            goto err; \
+        safes = PKCS12_unpack_authsafes(p12); \
+        if (!TEST_ptr(safes)) \
+            goto err;
+
+#define start_check_pkcs12_file(fname)  \
+    {  \
+        STACK_OF(PKCS7) *safes = NULL; \
+        int safe_idx = 0;              \
+        PKCS12 *p12 = read_p12(fname, NULL); \
+        if (!TEST_ptr(p12)) \
+            goto err; \
+        safes = PKCS12_unpack_authsafes(p12); \
+        if (!TEST_ptr(safes)) \
+            goto err;
```

Comment:
With the way all these macros work any goto err will cause a memory leak.

Reply:
Removed macros in later commit